### PR TITLE
Mmahlow

### PR DESCRIPTION
**Pull request by TG9541**
Contributions from M.Mahlow. Great Forth! Thanks a lot!

### CODE REVIEW

**Comment by TG9541** [comment]
Mr. Mahlow contributed some well crafted code, and he also proposed using the e4thcom `/res` feature for  creating bit/address constants for on-chip peripherals. WiP on my side.

**Comment by VK6TT** [comment]
Hi Thomas
 
I like the contribution by MM on STM8S103 Register ans Bit Identifiers. I am
confused though by the word "EQU". Is that a compiler directive or a Forth
word please?
 
Regards
Richard

  _____

**Comment by TG9541** [comment]
Hi Richard,
`equ` is a Forth word, but one that's interpreted by the Forth underneath e4thcom (which is based on MINFORTH), and like the directives `#include`, `#require`, or `\\ example` it's interpreted by e4thcom. 

Like a C preprocessor (and unlike "metacompilation") e4thcom works on the level of source code presented to the target:

`5005 equ PB_ODR` is converted into `$5005 CONSTANT PB_ODR` during upload. 

Of course, it's also possible to write a STM8EF file with the same definitions in plain text. One of the tasks will be implementing `\res` in other tools like loadserial.py and codeloadTCP.py.

Please note that e4thcom also supports assembler, and disassembler, plugins. That's much more like metacompilation.

**Comment by VK6TT** [comment]
Hi Thomas,
 
thank you for the explanation. I haven't used e4thcom which is where I got
confused. I'm not a fan of linux or windows so I simply stick to what I know
- windows XP. I code in notepad++ and send the file, or parts of a file
using copy and paste, via teraterm. 
 
I'm working on a library for the nRF24L01 transceivers for a small project
here. I'll use the CONSTANT word and post my contribute my code when
finished.
 
Many thanks
Richard
 

  _____

**Comment by TG9541** [comment]
Hi Richard, 
I can only encourage you to try using e4thcom on Linux. It almost can't be compared with using a plain terminal emulation program (10 times faster, much more convenient, much more powerful). A 10 year old "dumped" PC, and a smallish Linux distribution (e.g. Debian) is all you need. Alternatively you can use a $15 Raspberry Pi.
/Thomas

File: lib/CONSTANT [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    literal    \ 'EVAL in interpret mode
+    literal @  \ current 'EVAL
+    - if ( in compile mode ) [compile] literal then
+  ;RAM ALIAS docon ( -- x| )
+
+  NVM
+
+  : CONSTANT ( "name" x -- )
+    : compile docon , [compile] [ overt immediate ; 
+
```
Comment (Eelkhoorn) [inline lib/CONSTANT]: RAM

**Comment by Eelkhoorn** [comment]
Hi Thomas

I am new to docker so I had to set it up first. But I wasn't successful in running docker-sdcc, I think because I am playing on a Raspberry. I might try later on on a x86 machine.

I have a little remark about STM8S103efr: whats the use  of 16 bit bitmasks? Wouldn't BIT0 through BIT7 be sufficient?

**Comment by TG9541** [comment]
Docker images contain binaries, and on DockerHub x86-64 is the norm. There are repositories for ARM, but I don't know how to "cross-build" ARM Docker images. Creating docker images on the Pi is most likely a bit slow.

In STM8EF the "cell" size is 16bit, so bit0 through bit15 makes sense. There are even some STM8 registers that must be read, or stored in pairs (e.g. with 2C!).

You're right about `persist.fs`. It indeed needs `ULOCKF` and `LOCKF`. I'll have to re-enable it in `defconfig.inc`. I also expect `globconf.inc` to override `defconfig.inc`. That's strange...

**Comment by TG9541** [comment]
One should be careful with deleting branches - for some reason Travis-CI gets triggered again.

**Pull request merged**